File: program/src/instruction.rs
//! Instruction types

use borsh::{BorshDeserialize, BorshSerialize};
use solana_program::{
    instruction::{AccountMeta, Instruction},
    pubkey::Pubkey,
    sysvar,
};
use spl_token as token;

/// Instructions supported by the program
#[derive(Debug, BorshDeserialize, BorshSerialize, PartialEq)]
pub enum PoolInstruction {
    /// Initializes a new program
    Initialize,

    /// Swap tokens
    Swap {
        /// Amount of token IN
        amount_in: u64,
    },
}

/// Creates 'Initialize' instruction.
pub fn initialize(
    program_id: &Pubkey,
    pool: &Pubkey,
    authority: &Pubkey,
    bank_mint: &Pubkey,
    pool_mint: &Pubkey,
    bank: &Pubkey,
) -> Instruction {
    let accounts = vec![
        // General
        // Sys
        // Programs
        AccountMeta::new(*pool, true),
        AccountMeta::new_readonly(*authority, false),
        AccountMeta::new_readonly(*bank_mint, false),
        AccountMeta::new(*pool_mint, false),
        AccountMeta::new(*bank, false),
        AccountMeta::new_readonly(sysvar::rent::id(), false),
        AccountMeta::new_readonly(token::id(), false),
    ];

    Instruction::new_with_borsh(*program_id, &PoolInstruction::Initialize, accounts)
}

/// Creates 'Swap' instruction.
pub fn swap(
    program_id: &Pubkey,
    pool: &Pubkey,
    pool_authority: &Pubkey,
    user_transfer_authority: &Pubkey,
    pool_mint: &Pubkey,
    bank: &Pubkey,
    sender: &Pubkey,
    recipient: &Pubkey,
    amount_in: u64,
) -> Instruction {
    let accounts = vec![
        AccountMeta::new_readonly(*pool, false),
        AccountMeta::new_readonly(*pool_authority, false),
        AccountMeta::new_readonly(*user_transfer_authority, true),
        AccountMeta::new(*pool_mint, false),
        AccountMeta::new(*bank, false),
        AccountMeta::new(*sender, false),
        AccountMeta::new(*recipient, false),
        AccountMeta::new_readonly(token::id(), false),
    ];

    Instruction::new_with_borsh(*program_id, &PoolInstruction::Swap { amount_in }, accounts)
}

File: web/src/helpers/instruction.ts
import { TOKEN_PROGRAM_ID, u64 } from '@solana/spl-token'
import { PublicKey, SYSVAR_RENT_PUBKEY, TransactionInstruction } from '@solana/web3.js'
import { config } from '../config'
import { encodeData } from '../utils/solana'
import { PoolInsructionLayouts } from './layout'

export type InitializeParams = {
  pool: PublicKey
  authority: PublicKey
  bankMint: PublicKey
  poolMint: PublicKey
  bank: PublicKey
}
export const initialize = ({ pool, authority, bankMint, poolMint, bank }: InitializeParams) => {
  const data = encodeData(PoolInsructionLayouts.Initialize)

  return new TransactionInstruction({
    keys: [
      { pubkey: pool, isSigner: true, isWritable: true },
      { pubkey: authority, isSigner: false, isWritable: false },
      { pubkey: bankMint, isSigner: false, isWritable: false },
      { pubkey: poolMint, isSigner: false, isWritable: true },
      { pubkey: bank, isSigner: false, isWritable: true },
      { pubkey: SYSVAR_RENT_PUBKEY, isSigner: false, isWritable: false },
      { pubkey: TOKEN_PROGRAM_ID, isSigner: false, isWritable: false },
    ],
    programId: new PublicKey(config.programId),
    data,
  })
}

export type SwapParams = {
  pool: PublicKey
  poolAuthority: PublicKey
  userTransferAuthority: PublicKey
  poolMint: PublicKey
  bank: PublicKey
  sender: PublicKey
  recipient: PublicKey
  amountIn: number
}
export const swap = ({
  pool,
  poolAuthority,
  userTransferAuthority,
  poolMint,
  bank,
  sender,
  recipient,
  amountIn,
}: SwapParams) => {
  const data = encodeData(PoolInsructionLayouts.Swap, { amountIn: new u64(amountIn).toBuffer() })

  return new TransactionInstruction({
    keys: [
      { pubkey: pool, isSigner: false, isWritable: false },
      { pubkey: poolAuthority, isSigner: false, isWritable: false },
      { pubkey: userTransferAuthority, isSigner: true, isWritable: false },
      { pubkey: poolMint, isSigner: false, isWritable: true },
      { pubkey: bank, isSigner: false, isWritable: true },
      { pubkey: sender, isSigner: false, isWritable: true },
      { pubkey: recipient, isSigner: false, isWritable: true },
      { pubkey: TOKEN_PROGRAM_ID, isSigner: false, isWritable: false },
    ],
    programId: new PublicKey(config.programId),
    data,
  })
}
